scio-bigtable/src/main/java/com/spotify/scio/bigtable/BigtableDoFn.java

```diff
@@ -0,0 +1,301 @@
+    this.instanceId = UUID.randomUUID();
+    this.options = options;
+    this.cacheSupplier = cacheSupplier;
+    this.semaphore = new Semaphore(maxPendingRequests);
+  }
+
+  @Setup
+  public void setup() {
+    LOG.info("Setup for {}", this);
+    session.computeIfAbsent(instanceId, instanceId -> {
+      LOG.info("Creating BigtableSession with BigtableOptions {}", options);
+      try {
+        return new BigtableSession(options);
+      } catch (IOException e) {
+        e.printStackTrace();
+        LOG.error("Failed to create BigtableSession", e);
+        throw new RuntimeException("Failed to create BigtableSession", e);
+      }
+    });
+    cache.computeIfAbsent(instanceId, instanceId -> cacheSupplier.createCache());
+  }
+
+  @StartBundle
+  public void startBundle() {
+    LOG.info("Start bundle for {}", this);
+    futures.clear();
+    results.clear();
+    errors.clear();
+    requestCount = 0;
+    resultCount = 0;
+  }
+
+  @SuppressWarnings("unchecked")
+  @ProcessElement
+  public void processElement(ProcessContext c, BoundedWindow window) {
+    flush(r -> c.output(KV.of(r.input, r.output)));
+
+    final A input = c.element();
+    B cached = cacheSupplier.get(instanceId, input);
+    if (cached != null) {
+      c.output(KV.of(input, cached));
+      return;
+    }
+
+    final UUID uuid = UUID.randomUUID();
+    ListenableFuture<B> future;
+    try {
+      semaphore.acquire();
+      try {
+        future = asyncLookup(session.get(instanceId), input);
+      } catch (Exception e) {
+        semaphore.release();
+        LOG.error("Failed to process element", e);
+        throw e;
+      }
+    } catch (InterruptedException e) {
+      LOG.error("Failed to acquire semaphore", e);
+      throw new RuntimeException("Failed to acquire semaphore", e);
+    }
+    requestCount++;
+
+    // Handle failure
+    Futures.addCallback(future, new FutureCallback<B>() {
+      @Override
+      public void onSuccess(@Nullable B result) {
+        semaphore.release();
+      }
+
+      @Override
+      public void onFailure(Throwable t) {
+        semaphore.release();
+        errors.add(t);
+        futures.remove(uuid);
+      }
+    });
+
+    // Handle success
+    ListenableFuture<B> f = Futures.transform(future, new Function<B, B>() {
+      @Nullable
+      @Override
+      public B apply(@Nullable B output) {
+        cacheSupplier.put(instanceId, input, output);
+        results.add(new Result(input, output, c.timestamp(), window));
+        futures.remove(uuid);
+        return output;
+      }
+    });
+
+    // This `put` may happen after `remove` in the callbacks but it's OK since either the result
+    // or the error would've already been pushed to the corresponding queues and we are not losing
+    // data. `waitForFutures` in `finishBundle` blocks until all pending futures, including ones
+    // that may have already completed, and `startBundle` clears everything.
+    futures.put(uuid, f);
```

Comment:
Might this leak if this happens after `futures.remove(uuid)` ?

Reply:
Yes, `f` may be already complete before this happens. But we do `Futures.allAsList` in `finishBundle` to ensure all pending ones are handled and `futures.clear()` in `startBundle` to clean up. So there shouldn't be any leak? Is there a better way to handle this?